src/assign.rs

```diff
@@ -58,25 +54,456 @@ pub struct Assignment<'a> {
-    /// be `Some("/foo/bar")`.
-    pub created_or_mutated: PointerBuf,
-    /// A `Pointer` consisting of the path which was assigned.
+    /// replaced. For example, given the json `{ "foo:" "bar" }`, if `"new_value"` is
+    /// assigned to `"/foo/bar/baz"`, then `assigned_to` would be `/foo/bar"`
     ///
     /// ## Example
     /// ```rust
-    /// use serde_json::json;
-    /// use jsonptr::{Pointer, Assign};
+    /// # use serde_json::json;
+    /// # use jsonptr::{Pointer, Assign};
     /// let mut data = json!({ "foo": ["zero"] });
     /// let mut ptr = Pointer::from_static("/foo/-");
     /// let assignment = data.assign(&mut ptr, "one").unwrap();
     /// assert_eq!(assignment.assigned_to, Pointer::from_static("/foo/1"));
     /// ```
     pub assigned_to: PointerBuf,
+
+    /// The value that was replaced, if any.
+    pub replaced: Option<Value>,
+}
+
+pub(crate) fn assign_value<'v>(
+    mut ptr: &Pointer,
+    mut value: &'v mut Value,
+    mut src: Value,
+) -> Result<Assignment<'v>, AssignError> {
+    let mut offset = 0;
+    let mut buf = PointerBuf::with_capacity(ptr.as_str().len());
+
+    while let Some((token, tail)) = ptr.split_front() {
+        let tok_len = token.encoded().len();
+
+        let assigned = match value {
+            Value::Array(array) => assign_array(token, tail, buf, array, src, offset)?,
+            Value::Object(obj) => assign_object(token, tail, buf, obj, src)?,
+            _ => assign_scalar(token, ptr, buf, value, src)?,
+        };
+        match assigned {
+            Assigned::Done(assignment) => {
+                return Ok(assignment);
+            }
+            Assigned::Continue {
+                next_buf,
+                next_value,
+                same_src,
+            } => {
+                buf = next_buf;
+                src = same_src;
+                value = next_value;
+                ptr = tail;
+            }
+        }
+        offset += 1 + tok_len;
+    }
+
+    // Pointer is root, we can replace `dest` directly
+    let replaced = Some(core::mem::replace(value, src.into()));
+    Ok(Assignment {
+        assigned: value,
+        replaced,
+        assigned_to: buf,
+    })
+}
+
+/// Indicates error occurred during an assignment
+#[derive(Debug, PartialEq, Eq)]
+pub enum AssignError {
+    /// A `Token` within the `Pointer` failed to be parsed as an array index.
+    FailedToParseIndex {
+        /// Offset of the partial pointer starting with the invalid index.
+        offset: usize,
+        /// The source [`ParseIndexError`]
+        source: ParseIndexError,
+    },
+    /// An `Index` `Token` within the `Pointer` was out of bounds of the
+    /// target array.
+    OutOfBounds {
+        /// Offset of the partial pointer starting with the invalid index.
+        offset: usize,
+        /// The source [`OutOfBoundsError`]
+        source: OutOfBoundsError,
+    },
+}
+impl fmt::Display for AssignError {
```

Comment:
```suggestion
}

impl fmt::Display for AssignError {
```

Reply:
It'd be nice if rustfmt's [`blank_lines_lower_bound`](https://rust-lang.github.io/rustfmt/?version=v1.4.30&search=#blank_lines_lower_bound) had a configuration that didn't double space every line but instead just the end of blocks.